tools/switch.sh

```diff
@@ -0,0 +1,211 @@
+  ./switch.sh cmd fw-info | grep -A1 'Currently Running'                        # display currently running firmware
+EOF
+}
+
+# execute <fn<path>> <args...>
+execute() {
+    local FUNCTION=$1 ARGS=( "${@:2}" )
+
+    if [ "$(type -t "$FUNCTION")" != "function" ]; then
+        echo "$1 is not a function."
+        exit 1
+    fi
+
+    local SWITCHES=("/dev/switchtec0" "/dev/switchtec1")
+    for SWITCH in "${SWITCHES[@]}";
+    do
+        "$FUNCTION" "$SWITCH" "${ARGS[@]}"
+    done
+}
+
+displaySlotStatus() {
+    local SWITCH_NAME=$1
+    local PAX_ID
+
+    # Physical slot ids are set into the hardware. These are the mappings
+    declare -a PAX0_SlotFromPhysicalPort=(
+        [8]=8
+        [10]=7
+        [12]=15
+        [14]=16
+        [16]=17
+        [18]=18
+        [20]=14
+        [22]=13
+        [48]=12
+    )
+    declare -a PAX1_SlotFromPhysicalPort=(
+        [8]=4
+        [10]=5
+        [12]=6
+        [14]=2
+        [16]=1
+        [18]=9
+        [20]=10
+        [22]=11
+        [48]=3
+    )
+
+    # Make sure we can get the PAX ID
+    if [ ! "$(switchtec fabric gfms-dump "$SWITCH_NAME" | grep "^PAX ID:" | awk '{print $3}')" ]; then
+        exit $?
+    fi
+
+    PAX_ID=$(switchtec fabric gfms-dump "$SWITCH_NAME" | grep "^PAX ID:" | awk '{print $3}')
+
+    if ! [[ $PAX_ID =~ [[:digit:]]+ ]]; then
+        echo "$PAX_ID is not an integer"
+        exit 1
+    elif ! (( PAX_ID >= 0 && PAX_ID <= 1 )); then
+        echo "$PAX_ID not in range 0-1"
+        exit 1
+    fi
+
+    # Grab the attach status of each physical port
+    mapfile -t physicalPortStatus < <(switchtec fabric gfms-dump "$SWITCH" | grep " Physical Port ID")
+
+    local physicalPortString
+    for physicalPortString in "${physicalPortStatus[@]}";
+    do
+        local PHYSICAL_PORT_ID
+        PHYSICAL_PORT_ID=$(echo "$physicalPortString" | awk '{print $4}')
+        case $PAX_ID in
+            0)
+                SLOT=${PAX0_SlotFromPhysicalPort[$PHYSICAL_PORT_ID]}
+                ;;
+            1)
+                SLOT=${PAX1_SlotFromPhysicalPort[$PHYSICAL_PORT_ID]}
+                ;;
+            *)
+                exit 1
+        esac
+
+        printf "DEVICE: %s PAX_ID: %d SLOT: %d\t%s\n" "$SWITCH_NAME" "$PAX_ID" "${SLOT//}" "${physicalPortString//}"
+    done
+}
+
+alias TIME=""
+while getopts "th:" OPTION
+do
+    case "${OPTION}" in
+        't')
+            alias TIME=time
+            export TIMEFORMAT='%3lR'
+            ;;
+        'h',*)
+            usage
+            exit 0
+            ;;
+    esac
+done
+shift $((OPTIND - 1))
+
+if [ $# -lt 1 ]; then
+    usage
+    exit 1
+fi
+
+case $1 in
+    slot-info)
+        function slot-info() {
+            local SWITCH=$1
+            echo "Execute slot-info on $SWITCH"
+            TIME displaySlotStatus "$SWITCH"
+        }
+        execute slot-info
```

Comment:
Couldn't you do without the functions? Why do we need a function here?

Reply:
I'm emulating the style that Nate used in the `nvme.sh` script. Each function is called by the `execute` function which in this case iterates through all the switches calling the same function for each switch. I save rewriting the switch iteration code in each of these separate functions.
In this script it is not as useful as it was in `nvme.sh`, but if you are coming to this script from `nvme.sh` it should be familiar.

Reply:
Makes sense. Thanks for the clarification.